### ISSUE TITLE
Lint for if x.is_ok/some { x.unwrap() }

### ISSUE THREAD

**Issue by Roughsketch** [rust-lang-nursery/rust-clippy#1770]
I recently came across this in a project where I had the following code from when I was just starting to learn Rust. I found it by using pedantic mode and changing unwraps and was surprised it wasn't caught by default.

Here it is simplified:

```rust
if x.is_some() {
    let y = x.unwrap();
    // Use y
}
```
This can be better written as
```rust
if let Some(y) = x {
    // Use y
}
```

**Comment by oli-obk** [comment]
I am just as surprised we don't lint this yet. Unfortunately there might be false positives due to things not supported by the borrow checker. So we need to test this on a few codebases before adding it as warn by default.

**Comment by clarcharr** [comment]
Don't forget `is_err` and `unwrap_err` too.

**Comment by fanzier** [comment]
I'd like to try this.
